sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -78,6 +77,9 @@ class SparkOptimizer(
       PushPredicateThroughNonJoin,
       RemoveNoopOperators) :+
     Batch("User Provided Optimizers", fixedPoint, experimentalMethods.extraOptimizations: _*) :+
+    Batch("Merge Scalar Subqueries", Once,
```

Comment:
I've moved the `MergeScalarSubqueries` rule to the end of optimization phase, just before `ReplaceCTERefWithRepartition`. This is needed because we need to peek into the physical plans.